browser/brave_shields/brave_shields_util_profiles_unittest.cc

```diff
@@ -0,0 +1,212 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_shields/content/browser/brave_shields_util.h"
+#include "chrome/browser/content_settings/cookie_settings_factory.h"
+#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/test/base/testing_browser_process.h"
+#include "chrome/test/base/testing_profile.h"
+#include "chrome/test/base/testing_profile_manager.h"
+#include "components/content_settings/core/browser/cookie_settings.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "content/public/test/browser_task_environment.h"
+#include "testing/gtest/include/gtest/gtest.h"
+#include "url/gurl.h"
+
+namespace {
+
+using ScopedIncognitoProfile =
+    std::unique_ptr<Profile, decltype([](Profile* p) {
+                      ASSERT_TRUE(!!p->IsOffTheRecord());
+                      p->GetOriginalProfile()->DestroyOffTheRecordProfile(p);
+                    })>;
+
+testing::Message& operator<<(testing::Message& m,
+                             brave_shields::ControlType c) {
+  return m << brave_shields::ControlTypeToString(c);
+}
+
+}  // namespace
+
+namespace brave_shields {
+
+class BraveShieldsUtilProfilesTest : public testing::Test {
+ public:
+  BraveShieldsUtilProfilesTest()
+      : local_state_(TestingBrowserProcess::GetGlobal()) {}
+  ~BraveShieldsUtilProfilesTest() override = default;
+
+  TestingProfile* regular_profile() { return &profile_; }
+  ScopedIncognitoProfile incognito_profile() {
+    return ScopedIncognitoProfile(regular_profile()->GetOffTheRecordProfile(
+        Profile::OTRProfileID::PrimaryID(),
+        /*create_if_needed=*/true));
+  }
+
+  HostContentSettingsMap* hcsm(Profile* profile) {
+    return HostContentSettingsMapFactory::GetForProfile(profile);
+  }
+
+  HostContentSettingsMap* hcsm(ScopedIncognitoProfile& profile) {
+    return hcsm(profile.get());
+  }
+
+  const GURL test_url{"https://example.com"};
+
+  template <typename ValueT, typename SetFn, typename GetFn>
+  void RunTest(base::span<const std::pair<ValueT, ValueT>> cases,
+               SetFn setter,
+               GetFn getter) {
+    testing::Message issues;
+    for (const auto& [value, expect] : cases) {
+      setter(hcsm(regular_profile()), value);
+      const auto get = getter(hcsm(regular_profile()));
+      if (expect != get) {
+        issues << "Regular profile: Set " << value << " Get " << get
+               << " Expected " << expect << "\n";
+      }
+
+      auto incognito = incognito_profile();
+      // Now change value for incognito and expect that values are not depended
+      // on the regular profile.
+      for (const auto& [ivalue, iexpect] : cases) {
+        setter(hcsm(incognito), ivalue);
+        const auto iget = getter(hcsm(incognito));
```

Comment:
Nit: it looks like we are repeating code here hen we could be calling `RunTest` once for the normal profile, and again for the incognito profile, by passing the pointer to the profile. I'm not sure which is best, but just thought of mentioning.